Core/Epaper/Src/epaper.c

```diff
@@ -0,0 +1,462 @@
+		if (!(xSemaphoreTake(epdSpiSemHandle, 10) == pdTRUE))
+		{
+			Debug("Interrupt failed to occur during EPAP SPI transmit\n");
+		}
+
+		// Set Epaper to data mode
+		setDataMode();
+
+		// Write command to spi
+		HAL_SPI_Transmit_IT(&hspi2, data, numBytes);
+		// Wait for spi interupt signaling data transfer complete - timout 50ms
+		if (!(xSemaphoreTake(epdSpiSemHandle, 50) == pdTRUE))
+		{
+			Debug("Interrupt failed to occur during EPAP SPI transmit\n");
+		}
+
+		// Close spi communication
+		csOff();
+}
+
+/*!
+  @brief   Set Epaper internal lookup table and voltage data
+  @param   lut 	Look up table to write to epaper
+*/
+static void setLookupTable(uint8_t *lut)
+{
+	// Set look up table register - last 6 bytes belong to seperate registers
+	sendMessage(CMD_WRITE_LUT_REGISTER, lut, LUT_SIZE - 6);
+	// Wait for busy_release interupt
+	waitBusyRelease(); 
+
+	// Set Mystery Register 1
+	sendBuffer[0] = lut[153];		
+	sendMessage(CMD_WRITE_MYSTERY_REGISTER_2, sendBuffer, 1);
+
+	// Set Gate voltage Control Register
+	sendBuffer[0] = lut[154];		
+	sendMessage(CMD_GATE_DRIVING_VOLTAGE_CONTROL, sendBuffer, 1);
+
+	// Set Source Driving Voltage Register
+	sendBuffer[0] = lut[155];
+	sendBuffer[1] = lut[156];
+	sendBuffer[2] = lut[157];
+	sendMessage(CMD_SOURCE_DRIVING_VOLTAGE_CONTROL, sendBuffer, 3);
+
+	// Set VCOM Register
+	sendBuffer[0] = lut[158];
+	sendMessage(CMD_WRITE_VCOM_REGISTER, sendBuffer, 1);
+}
+
+/*!
+  @brief   Turn on Epaper display
+*/
+static void turnOnDisplay()
+{
+	// Activate display update sequence
+	sendCommand(CMD_MASTER_ACTIVATION);
+
+	// Wait for busy_release interupt
+	waitBusyRelease();
+}
+
+/*!
+  @brief	Set Window Size of epaper
+  @param	Xstart	Starting X coordinate
+  @param	Ystart	Starting Y coordinate
+  @param	Xend 	Ending X coordinate
+  @param	Yend	Ending Y coordinate
+*/
+static void setWindows(uint16_t Xstart, uint16_t Ystart, uint16_t Xend, uint16_t Yend)
+{
+	// SET_RAM_X_ADDRESS_START_END_POSITION
+	sendBuffer[0] = (Xstart>>3) & 0xFF;
+	sendBuffer[1] = (Xend>>3) & 0xFF;
+	sendMessage(CMD_SET_RAM_X_WINDOW, sendBuffer, 2);
+	
+	// SET_RAM_Y_ADDRESS_START_END_POSITION
+	sendBuffer[0] = Ystart & 0xFF;
+	sendBuffer[1] = (Ystart >> 8) & 0xFF;
+	sendBuffer[2] = Yend & 0xFF;
+	sendBuffer[3] = (Yend >> 8) & 0xFF;
+    sendMessage(CMD_SET_RAM_Y_WINDOW, sendBuffer, 4);
+}
+
+/*!
+  @brief	Set Cursor position of epaper
+  @param	Xstart	Starting X coordinate
+  @param	Ystart	Starting Y coordinate
+*/
+static void setCursor(uint16_t Xstart, uint16_t Ystart)
+{
+	// SET_RAM_X_ADDRESS_COUNTER
+	sendBuffer[0] = (Xstart & 0xFF);
+	sendMessage(CMD_SET_RAM_X_COUNTER, sendBuffer, 1);
+
+	// SET_RAM_Y_ADDRESS_COUNTER
+	sendBuffer[0] = Ystart & 0xFF;
+	sendBuffer[1] = ((Ystart >> 8) & 0xFF);
+    sendMessage(CMD_SET_RAM_Y_COUNTER, sendBuffer, 2);
+}
+
+/*!
+  @brief	Initialize epaper display
+*/
+static void setFullRefreshSettings()
+{
+	// Set initial pin configuration
```

Comment:
Do we potentially want to check that epdConfig is not already `FULL_REFRESH`? Only a good idea if you think it's likely that setFullRefreshSettings() will be called multiple times in a row without switching modes